src/_image_resample.h

```diff
@@ -500,240 +500,58 @@ typedef enum {
 } interpolation_e;
 
 
-template <typename T>
-class type_mapping;
+// T is rgba iff it has an T::r field.
```

Comment:
```suggestion
// T is rgba if it has an T::r field.
```

Reply:
Isn't this iff = if and only if?

Reply:
Evidently that wasn't clear to me.

Maybe it should be written out in full? The templated code is essentially obfuscated, so maybe the English should try to go in the other direction.

Reply:
Is there more that can be done to clarify the code, via some combination of comments and changes in the code itself?

Reply:
No, not really. Clever templated C++ is usually difficult to understand unless you are seeing it regularly.

This is a significant reduction in the number of lines of code. Each line is more difficult to understand, but the reduction in the number of lines means it is probably easier to understand overall.

Reply:
> Evidently that wasn't clear to me.

Good point.

Reply:
Sure, wrote out "if and only if".

Reply:
@ianthomas23 Do you think writing type_mapping as
```c++
template<typename color_type, typename Enable = void>
struct type_mapping {};


template<typename color_type>
struct type_mapping<color_type, typename std::enable_if<!is_grayscale<color_type>::value>::type>
{
    using blender_type = typename std::conditional<
        std::is_same<color_type, agg::rgba8>::value,
        fixed_blender_rgba_plain<color_type, agg::order_rgba>,
        agg::blender_rgba_plain<color_type, agg::order_rgba>
    >::type;
    using pixfmt_type =
        agg::pixfmt_alpha_blend_rgba<blender_type, agg::rendering_buffer>;
    using pixfmt_pre_type = typename
        agg::pixfmt_alpha_blend_rgba<
            typename std::conditional<
                std::is_same<color_type, agg::rgba8>::value,
                fixed_blender_rgba_pre<color_type, agg::order_rgba>,
                agg::blender_rgba_pre<color_type, agg::order_rgba>
            >::type,
            agg::rendering_buffer>;
    template<typename A> using span_gen_affine_type =
        agg::span_image_resample_rgba_affine<A>;
    template<typename A, typename B> using span_gen_filter_type =
        agg::span_image_filter_rgba<A, B>;
    template<typename A, typename B> using span_gen_nn_type =
        agg::span_image_filter_rgba_nn<A, B>;
};


template<typename color_type>
struct type_mapping<color_type, typename std::enable_if<is_grayscale<color_type>::value>::type>
{
    using blender_type = agg::blender_gray<color_type>;
    using pixfmt_type = agg::pixfmt_alpha_blend_gray<blender_type, agg::rendering_buffer>;
    using pixfmt_pre_type = pixfmt_type;
    template<typename A> using span_gen_affine_type =
        agg::span_image_resample_gray_affine<A>;
    template<typename A, typename B> using span_gen_filter_type =
        agg::span_image_filter_gray<A, B>;
    template<typename A, typename B> using span_gen_nn_type =
        agg::span_image_filter_gray_nn<A, B>;
};
```
(essentially getting rid of most std::conditional in favor of partial template specialization) reads better?

Reply:
Very slightly. But as there is already use of `std::conditional` maybe it is better to keep all of them for consistency. On balance, I'd be inclined to keep it as it is (commit `9644a9c`).

Reply:
(Note that I could also split out the agg::rgba8 case as another template specialization (duplicating some of the rgba type_mapping) to fully get rid of std::conditional, if you prefer.)